### PR TITLE
Fix `tox_root` propagation to `work_dir`

### PR DESCRIPTION
**Pull request by Tbruno25**
This applies a fix for issue #2933 where `work_dir` was not accurately including the current `tox_root` during test runs

# Thanks for contribution

Please, make sure you address all the checklists (for details on how see
[development documentation](http://tox.readthedocs.org/en/latest/development.html#development))!

- [x] ran the linter to address style issues (`tox -e fix`)
- [x] wrote descriptive pull request text
- [x] ensured there are test(s) validating the fix
- [x] added news fragment in `docs/changelog` folder
- [ ] updated/extended the documentation

### CODE REVIEW

**Review by gaborbernat — CHANGES_REQUESTED** [review]
CI is failing.

**Comment by Tbruno25** [comment]
> Let me know if you'd like me to open a PR onto your branch, to incorporate these changes.

Let's merge it in!
 
>   I may be wrong, but my understanding of the control flow is that the `conf.work_dir` can never actually be `None`. In `Config.make`, it's always set to a non-`None` value:
> ...
>   * The default case gets handled by returning `case(Path, self["tox_root"])) / ".tox"`
>   * The case when `work_dir` is set by `--workdir` is handled here:

Agree with your conclusion.

>   https://github.com/tox-dev/tox/blob/a4be5d8720d2976a9e8b59ee2c0379db355b60ea/src/tox/config/sets.py#L196-L197
> * I wrote tests in `tests/test_run.py` (instead of `tests/test_main.py`), since there was already a test for `--workdir` there which would've been similar to the test of `--root`.
> * I believe I have some fixes for some of your test failures

I was _really_ hoping to avoid altering existing tests which is why I was "guess and checking" with the different fixes. If you've already re-worked these then even more reason to pull it in :+1:

**Comment by Tbruno25** [comment]
> CI is failing.

Should be all set thanks to @kdestin :rocket:

File: src/tox/config/sets.py [hunk 1]
```diff
@@ -188,23 +188,17 @@ def register_config(self) -> None:
...
-
-        def work_dir_post_process(value: Path) -> Path:
-            return self._conf.work_dir if self._conf.options.work_dir else value
-
         self.add_config(
             keys=["work_dir", "toxworkdir"],
             of_type=Path,
-            default=work_dir_builder,
-            post_process=work_dir_post_process,
+            default=lambda *_: self["tox_root"] / ".tox",  # noqa: U101
```
Comment (gaborbernat) [inline src/tox/config/sets.py:194]: Why are we using here the * operator and not enumerating the argumetns? Why is this now a lambda and not a method as before? I prefered the previous way.

File: src/tox/config/sets.py [hunk 2]
```diff
@@ -188,23 +188,17 @@ def register_config(self) -> None:
...
+            post_process=lambda dir: self._conf.work_dir if self._conf.options.work_dir else dir,
             desc="working directory",
         )
         self.add_config(
             keys=["temp_dir"],
             of_type=Path,
-            default=lambda conf, _: cast(Path, self["work_dir"]) / ".tmp",  # noqa: U100, U101
+            default=lambda *_: self["work_dir"] / ".tmp",  # noqa: U101
```
Comment (gaborbernat) [inline src/tox/config/sets.py:201]: Same.

File: tests/tox_env/python/test_python_runner.py [hunk 3]
```diff
@@ -95,7 +95,16 @@ def test_journal_package_dir(tmp_path: Path) -> None:
 
 def test_package_temp_dir_view(tox_project: ToxProjectCreator, demo_pkg_inline: Path) -> None:
     project = tox_project({"tox.ini": "[testenv]\npackage=wheel"})
-    result = project.run("r", "-vv", "-e", "py", "--root", str(demo_pkg_inline))
+    result = project.run(
```
Comment (gaborbernat) [inline tests/tox_env/python/test_python_runner.py:98]: This seems an unrelated test, why did we change this?

**Review by gaborbernat — CHANGES_REQUESTED** [review]

File: src/tox/config/sets.py (hunk 1 above)
Comment (Tbruno25) [reply src/tox/config/sets.py:194]: The args have no impact on what the function will return. I opted for the star as it allowed a single `noqa` code instead of multiple. 

And a lambda was utilized in the  `temp_dir` config already. I have no strong opinion on lambda/funcs but I think the implementation should be consistent either way.

Will revert!

File: tests/tox_env/python/test_python_runner.py (hunk 3 above)
Comment (Tbruno25) [reply tests/tox_env/python/test_python_runner.py:98]: @kdestin may be able to speak to this better than I

Comment (kdestin) [reply tests/tox_env/python/test_python_runner.py:98]: This test seems to have depended on the bug we're fixing. Adding `"--workdir", str(project.path / ".tox")` preserves the expected state of the tox invocation.

File: src/tox/config/sets.py [hunk 4]
```diff
@@ -179,6 +179,15 @@ def __init__(self, conf: Config, section: Section, root: Path, src_path: Path) -
         desc = "define environments to automatically run"
         self.add_config(keys=["env_list", "envlist"], of_type=EnvList, default=EnvList([]), desc=desc)
 
+    def _default_work_dir(self, conf: Config, env_name: str | None) -> Path:  # noqa: U100
+        return cast(Path, self["tox_root"] / ".tox")
+
+    def _default_temp_dir(self, conf: Config, env_name: str | None) -> Path:  # noqa: U100
+        return cast(Path, self["work_dir"] / ".tmp")
+
+    def _work_dir_post_process(self, dir: Path) -> Path:
```
Comment (gaborbernat) [inline src/tox/config/sets.py:188]: Why move these to class methods from before being inlined methods?

File: docs/changelog/2933.bugfix.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+Fix issue where ``work_dir`` was not correctly including ``tox_root`` for test runs
```
Comment (gaborbernat) [inline docs/changelog/2933.bugfix.rst:1]: trailing period, to finish your sentence and be grammatically correct

File: src/tox/config/sets.py (hunk 4 above)
Comment (Tbruno25) [reply src/tox/config/sets.py:188]: I tend to try and avoid nested functions for a variety of reasons. By defining them as methods it improves testability while simultaneously making `register_config` shorter and easier to digest at first glance.

It's more idiomatic imo -- of course I'm happy to revert if you feel differently!

**Review by gaborbernat — APPROVED** [review]
